mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -289,24 +301,20 @@ class Word2Vec extends Serializable with Logging {
     val expTable = sc.broadcast(createExpTable())
     val bcVocab = sc.broadcast(vocab)
     val bcVocabHash = sc.broadcast(vocabHash)
-
-    val sentences: RDD[Array[Int]] = words.mapPartitions { iter =>
-      new Iterator[Array[Int]] {
-        def hasNext: Boolean = iter.hasNext
-
-        def next(): Array[Int] = {
-          val sentence = ArrayBuilder.make[Int]
-          var sentenceLength = 0
-          while (iter.hasNext && sentenceLength < MAX_SENTENCE_LENGTH) {
-            val word = bcVocabHash.value.get(iter.next())
-            word match {
-              case Some(w) =>
-                sentence += w
-                sentenceLength += 1
-              case None =>
-            }
+    // each partition is a collection of sentences,
+    // will be translated into arrays of Index integer
+    val sentences: RDD[Array[Int]] = dataset.mapPartitions { sentenceIter =>
+      // Each sentence will map to 0 or more Array[Int]
+      sentenceIter.flatMap { sentence => {
+          // Sentence of words, some of which map to a word index
+          val wordIndexes = sentence.flatMap(bcVocabHash.value.get)
+          if (wordIndexes.nonEmpty) {
```

Comment:
They should be equivalent Scala code. We don't need to mix in their meanings. Let us try providing an example of  `sentenceIter: Iterator[Array[String]]` and `vocabHash: Map[String, Int]` such that

~~~scala     
    sentenceIter.flatMap { sentence =>
        val wordIndexes = sentence.flatMap(vocabHash.get)
        wordIndexes.grouped(maxSentenceLength).map(_.toArray)
    }
~~~~

returns a different result from your code:

~~~scala
      sentenceIter.flatMap { sentence =>
        val wordIndexes = sentence.flatMap(vocabHash.get)
        if (wordIndexes.nonEmpty) {
          val sentenceSplit = wordIndexes.grouped(maxSentenceLength)
          sentenceSplit.map(_.toArray)
        } else {
          None
        }
     }
~~~

Essentially we are comparing the behavior when `wordIndexes` is an empty `Iterator[Int]`. In this case, no matter what value `maxSentenceLength` takes, ` wordIndexes.grouped(maxSentenceLength).map(_.toArray)` returns an empty iterator, which will be skipped by `sentenceIter.flatMap`. So it is the same as `None` being skipped by `sentenceIter.flatMap` in the current implementation.